apps/arweave/src/ar_webhook.erl

```diff
@@ -239,16 +300,306 @@ do_call_webhook(URL, Headers, Entity, Event, _N) ->
 
-to_json(#block {} = Block) ->
+to_json(#block{} = Block) ->
 	{JSONKVPairs} = ar_serialize:block_to_json_struct(Block),
 	JSONStruct = {lists:keydelete(wallet_list, 1, JSONKVPairs)},
 	ar_serialize:jsonify({[{block, JSONStruct}]});
-to_json(#tx {} = TX) ->
+to_json(#tx{} = TX) ->
 	{JSONKVPairs1} = ar_serialize:tx_to_json_struct(TX),
 	JSONKVPairs2 = lists:keydelete(data, 1, JSONKVPairs1),
 	JSONKVPairs3 = [{data_size, TX#tx.data_size} | JSONKVPairs2],
 	JSONStruct = {JSONKVPairs3},
-	ar_serialize:jsonify({[{transaction, JSONStruct}]}).
+	ar_serialize:jsonify({[{transaction, JSONStruct}]});
+to_json(Map) when is_map(Map) ->
+	jiffy:encode(Map).
+
+handle_sync_record_add_range(Start, End, StoreID, State) ->
+	handle_sync_record_add_range(Start, End, StoreID, State, 0).
+
+handle_sync_record_add_range(Start, End, StoreID, State, N) when N < ?NUMBER_OF_TRIES ->
+	case get_tx_offset_data(Start, End, State#state.tx_offset_cache) of
+		{ok, Data, Cache} ->
+			process_updated_tx_data(Data, StoreID, State#state{ tx_offset_cache = Cache });
+		not_found ->
+			State;
+		Error ->
+			?LOG_WARNING([{event, failed_to_process_webhook_sync_record_add_range},
+					{error, io_lib:format("~p", [Error])},
+					{range_start, Start},
+					{range_end, End}]),
+			timer:sleep(?WAIT_BETWEEN_TRIES),
+			handle_sync_record_add_range(Start, End, State, N + 1)
+	end;
+handle_sync_record_add_range(Start, End, _StoreID, State, _N) ->
+	?LOG_WARNING([{event, gave_up_webhook_tx_offset_data_fetch},
+		{range_start, Start},
+		{range_end, End},
+		{number_of_tries, ?NUMBER_OF_TRIES},
+		{wait_between_tries, ?WAIT_BETWEEN_TRIES}
+	]),
+	State.
+
+process_updated_tx_data([], _StoreID, State) ->
+	State;
+process_updated_tx_data([{TXID, Start, End} | Data], StoreID, State) ->
+	Cache = State#state.tx_offset_cache,
+	case cache_is_tx_data_marked_synced(TXID, Cache) of
+		true ->
+			process_updated_tx_data(Data, StoreID, State);
+		false ->
+			process_updated_tx_data(Data, StoreID,
+					maybe_call_transaction_data_synced(Start, End, TXID, StoreID, State))
+	end.
+
+%% @doc Return a list of {TXID, Start2, End2} for recorded transactions from the given range.
+get_tx_offset_data(Start, End, Cache) ->
+	case cache_get_tx_offset_data(Start, End, Cache) of
+		[] ->
+			case ar_data_sync:get_tx_offset_data_in_range(Start, End) of
+				{ok, []} ->
+					not_found;
+				{ok, Data} ->
+					Cache2 = cache_add_tx_offset_data(Data, Cache),
+					{ok, Data, Cache2};
+				Error ->
+					Error
+			end;
+		List ->
+			{ok, List, Cache}
+	end.
+
+cache_get_tx_offset_data(Start, End, Cache) ->
+	#tx_offset_cache{
+		start_offset_txid_end_offset = Set
+	} = Cache,
+	Iterator = gb_sets:iterator_from({Start - 1, infinity, infinity}, Set),
+	cache_get_tx_offset_data_from_iterator(End, Iterator).
+
+cache_get_tx_offset_data_from_iterator(End, Iterator) ->
+	case gb_sets:next(Iterator) of
+		none ->
+			[];
+		{{Start2, TXID, End2}, Iterator2} when Start2 < End ->
```

Comment:
This function includes all ranges for which Start is less than End (i.e. it may include ranges for which End2 is greater than End)

By my read the [get_tx_offset_data_in_range](https://github.com/ArweaveTeam/arweave/pull/556/files#diff-2934ddad2ed46ac194a6be7abde4be6d440bc7565e1128bba234fad9fd2bd5d2R1668) function only includes ranges where both Start and End are within the queried Range.

Just confirming that I've understood the differing logic correctly and that this difference is expected/fine.

Reply:
Updated it [here](https://github.com/ArweaveTeam/arweave/pull/556/commits/eafad4b57c915808f2a7948d54d943e5315971ee) - changed from {Start, TXID, End} to {End, TXID, Start} because gb_sets does not have an ar_kv:get_prev analogue; the logic should be the same across two places now.